YAML test framework: SkipSection refactoring

ldematte opened the pull request:
This PR is a preliminary refactoring step to make the introduction of feature-based skip criteria easier to add and test

Introduce `SkipVersionContext` and consolidate skip() methods
Introduce `SkipCriteria` for modular skip decisions
Removed accessors to specific criteria on `SkipSection` (used only on tests), adjusted tests and tests assertions:
- separate `VersionRangeTests` to test parsing version ranges
- parse tests are now on `SkipVersionBuilder`, so we inspect results from parsing _before_ building the final SkipSection evaluator
- the final SkipSection is tested directly on the effects of `skip`; added more of these tests to cover different combinations (features, versions, OSes, etc.)

ldematte committed:
Introduce SkipVersionContext and consolidate skip() methods

ldematte committed:
Merge remote-tracking branch 'upstream/main' into refactoring-yaml-skipsection

ldematte committed:
Intermediate step: moved section construction to a builder

ldematte committed:
Introduce SkipCriteria for modular skip decisions

- Refined SkipSectionContext to uniform boolean tests on the cluster properties
- Removed accessors to specific criteria from SkipSection (used only on tests), adjusted test assertions
- Version range checks commented out, as they have been already addressed in a pending PR (#103311)

ldematte committed:
spotless

ldematte committed:
Refactoring tests

- moving and adding VersionRange tests
- adding specific version and os skip tests
- modified parse/validate/build to make SkipSection more unit-testable

ldematte committed:
spotless

ldematte committed:
Update external parse tests

ldematte committed:
Separate and refine skip tests

elasticsearchmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

ldematte committed:
Merge remote-tracking branch 'upstream/main' into refactoring-yaml-skipsection

ldematte committed:
spotless

mosche commented on test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/ESClientYamlSuiteTestCase.java:
```diff
@@ -612,4 +610,28 @@ protected final RestClientBuilder getClientBuilderWithSniffedHosts() throws IOEx
         return testCandidate;
     }
+
+    private static class SkipSectionContextAdapter implements SkipSectionContext {
```
I'm wondering, is `SkipSectionContext` really necessary? Imho things are becoming just harder to understand due to the added level of indirection. As far as I can see this is only beneficial for testing. Wouldn't it be possible to just mock `ClientYamlTestExecutionContext` to achieve the same and use `skip(ClientYamlTestExecutionContext)` instead?

ldematte replied:
I prefer (and find easier to understand) when things are more scoped and dependencies have clear cut boundaries, and then add adapters when needed. Besides making this easier to test (as you noticed), it makes it clearer what Skip (SkipSection/SkipCriteria) uses and what not. 
That said, since these classes are all specific to section execution in Yaml tests, I see not much as a problem to get rid of SkipSectionContext and use ClientYamlTestExecutionContext directly. I can go in that direction.

mosche replied:
> it makes it clearer what Skip (SkipSection/SkipCriteria) uses and what not

Absolutely on board with that part, but in that case I would prefer if SkipSectionContext was just a subset of the broader context (os, clusterHasFeature, esVersion). You're still implementing your skip logic on the full scope of ClientYamlTestExecutionContext, and the tricky / twisted bit is that the checks are not consistent with respect to skipping (clusterIsRunningOs / clusterVersionInRange vs clusterHasFeature). But consider it a nit.

ldematte replied:
Yes, I agree they are identical now (same set of things exposed) but with different logic. Since there is very little utility for `SkipSectionContext` as is, I will remove it.

mosche reviewed (commented):
@Idematte I might be missing some context with respect to upcoming changes here. At the moment I'm slightly hesitant, pls have a look at my comments below.

ldematte commented on test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/SkipCriteria.java:
```diff
@@ -0,0 +1,13 @@
+
+package org.elasticsearch.test.rest.yaml.section;
+
+public interface SkipCriteria {
```
I think the skip decision is a good idea, but I kept `testFeatures` separate for another reason; test features are "static" meta features supported by the test runner, not the cluster. It's a semi-fixed (mostly at compile time) set of "things" that the test runner (not the cluster under examination) must support, or the test is not run. That's why I wanted to keep this separate.
(Also, I suspect this was more used in the past, and now it's mainly used to test xpack/no_xpack).

Maybe I should rename this as to include `Cluster` somewhere? `ClusterSkipCriteria`?

ldematte replied:
Re: keeping it as is, the reason I did this PR is that I was introducing a skip criteria based on features, and cramming it into the current framework was definitely not "clean"; you can already see the thing was designed with just version of test features in mind, and OS was retrofitted later.
I wanted to avoid adding another separate check (with unclear relationships with the existing ones - does it run before or after?) and make things more explicit, and make this change so now there is one `skip` check which does all the checks, built incrementally with composable criteria from parsing.

mosche replied:
Oh, I get it... this part isn't related to the feature api at all. Sorry, that was confusing but makes sense now. How about evaluating the testFeature check in the builder to make that clear? Your builder could return an overridden `SkipAlwaysSection` in that case while `SkipSection` is only based on skip criteria?

ldematte replied:
Unfortunately SkipSection already exposes to some extent test features (for cross checking/validation). We can work to make that better, but I think it's for another PR. Meanwhile I reduced it to just a test method (yamlRunnerHasFeature), instead of esposing the whole list, and still used your idea to inject a single always-true skip criteria.

thecoop commented on test/yaml-rest-runner/src/test/java/org/elasticsearch/test/rest/yaml/section/AbstractClientYamlTestFragmentParserTestCase.java:
```diff
@@ -39,4 +40,63 @@ public void tearDown() throws Exception {
+    protected static SkipSectionContext neverSkipContext() {
+        return new SkipSectionContext() {
+            @Override
+            public boolean clusterIsRunningOs(String osName) {
```
Couldn't these all be default implementations?

ldematte replied:
Yes, I thought about that. Not a big fan of default interface implementations, but sure, it can be done that way.

ldematte replied:
After removing SkipSectionContext and using mocks as @mosche suggested, this is no longer needed

thecoop commented on test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/SkipCriteria.java:
```diff
@@ -0,0 +1,13 @@
+
+package org.elasticsearch.test.rest.yaml.section;
+
+public interface SkipCriteria {
```
Isn't this just a `Predicate<SkipSectionContext>`?

ldematte replied:
Yes, that's true. I thought having an explicit interface was cleaner/nicer. I think it's weird to `implement Predicate<>`.
But I'm OK with `Predicate<SkipSectionContext>` + some sort of factory, e.g. `Predicate<SkipSectionContext> createVersionBasedSkipCriteria(args)` to replace the concrete classes and constructors. Wdyt?

thecoop replied:
That's a good solution, it keeps the semantics clear whilst making it easy to create implementations

mosche replied:
I was initially thinking the same, but then `SkipCriteria` is a bit more self-explanatory than `Predicate<SkipSectionContext>`. But how about moving the two impls into the interface?

```
    static SkipCriteria skipOs(List<String> operatingSystems){
        return context -> operatingSystems.stream().anyMatch(context::clusterIsRunningOs);
    }
```

thecoop commented on test/yaml-rest-runner/src/test/java/org/elasticsearch/test/rest/yaml/section/SkipSectionTests.java:
```diff
@@ -76,49 +113,33 @@ public void testParseSkipSectionVersionNoFeature() throws Exception {
+        assertThat(skipSectionBuilder, notNullValue());
+        assertThat(skipSectionBuilder.version, emptyOrNullString());
+        assertThat(skipSectionBuilder.testFeatures, hasSize(1));
+        assertThat(skipSectionBuilder.testFeatures, contains("regex"));
```
nit - `contains` checks the collection _only_ contains those items, and no others, so the size check above it is redundant

ldematte replied:
Good catch, will simplify it. Thanks!

ldematte committed:
Merge remote-tracking branch 'upstream/main' into refactoring-yaml-skipsection

ldematte committed:
Simplify

- Remove SkipSectionContext
- Replace SkipCriteria interface with Predicate + factory
- Moved Features check (YAML test runner features) to SkipSection build time

ldematte committed:
Merge remote-tracking branch 'upstream/main' into refactoring-yaml-skipsection

ldematte committed:
Separated check for xpack/no_xpack

Check for xpack is cluster-configuration (modules installed) dependent, while Features are meant to be "static" test-runner capabilities. We separate them so checks on one (test-runner features) can be run before and separately from the other.

ldematte committed:
Merge remote-tracking branch 'upstream/main' into refactoring-yaml-skipsection

ldematte committed:
Fix Version/Strings nodeVersions mismatch after merge

mosche commented:
Great progress on this, I like where this is going  :)

ldematte commented on test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/SkipSection.java:
```diff
@@ -54,22 +163,21 @@ public static SkipSection parse(XContentParser parser) throws IOException {
                 } else if ("reason".equals(currentFieldName)) {
-                    reason = parser.text();
+                    builder.withReason(parser.text());
                 } else if ("features".equals(currentFieldName)) {
```
Good question, I'm not sure about that myself.
On the one hand, there will be a strong risk of confusion.
On the other hand, this has been there for a lot of time - people might be used to it. 
With cluster features, we will need to probably add a couple of new fields, like `cluster_has_features` and `cluster_lacks_features` - remember, this will be under `skip`, and we probably want to skip mostly based on the fact that a cluster has not some feature. Wdyt?

mosche replied:
I wish that would be a `requires` instead of a negative `skip` section. That typically leads to inconsistent semantics, e.g. version (skipped if matching) vs. features (skipped if not matching). But anyways ...

Thinking about `cluster_has_features`, do you think there's use cases where a test should be skipped if a cluster has a feature, but it's expected to be absent? Almost seems like an anti pattern to me, but I'm probably being wishful here.

Considering `features` is already using inverted semantics, it's probably fine to have `cluster_features`. But in that case we better rename `features` to `test_runner_features` or similar I think...

ldematte replied:
I think we can (and probably should) have both `cluster_has_features` and `cluster_lacks_features` (or some better name I'm struggling with).
I have a PR ready (will push them in a moment) with both of them. There are definitely cases for skipping based on features present - it's the mirror of today. You want to split tests so that one runs before a change and one after (e.g. a change in response that needs different assertions). Today this is done with version ranges (e.g. [ - 8.11.99] and [8.12.0 - ]).

Yes, the double negative semantics gives headaches :) I ended up renaming things 2 or 3 times in the next PR.

ldematte commented on test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/SkipCriteria.java:
```diff
@@ -0,0 +1,67 @@
+            }
+
+            @Override
+            public String toString() {
```
I like the matcher, looks elegant :)
But I'm not sure this is the place: I've done that at some point and then reverted it.
Testing the effect of the skip predicate is already covered in `SkipSectionTests`, were multiple versions are tried to assert what is expected to be in range and what not. Seems like a duplication to do it elsewhere, where the focus is on "am I parsing each section correctly". But I agree this feels unbalanced and not ideal - I think we might just remove them. Like I said, parsing and effect are already covered elsewhere (`SkipSectionTests`), and if we feel we need more tests we should add them there.

mosche replied:
In that case I'm all on for removing these assertions 👍

ldematte replied:
Done, and removed the toString overload (so now version criteria does not stand out as "special")

ldematte committed:
Addressing comments

- removed not-so-helpful toString for Version SkipCriteria
- renaming
- removed un-needed file

ldematte committed:
Merge remote-tracking branch 'upstream/main' into refactoring-yaml-skipsection

ldematte commented:
I have https://github.com/elastic/elasticsearch-serverless/pull/1307 to address the serverless compatibility issue

ldematte commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into refactoring-yaml-skipsection

ldematte committed:
Added missing overload (we temporarily need it for serverless)

ldematte committed:
Merge branch 'refactoring-yaml-skipsection' of github.com:ldematte/elasticsearch into refactoring-yaml-skipsection

ldematte committed:
remove final

ldematte committed:
Merge remote-tracking branch 'upstream/main' into refactoring-yaml-skipsection

mosche commented on test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/SkipCriteria.java:
```diff
@@ -0,0 +1,58 @@
+        assert versionRanges.isEmpty() == false;
+        return context -> {
+            // Try to extract the minimum node version. Assume CURRENT if nodes have non-semantic versions
+            // TODO: push this logic down to VersionRange.
```
Do you want to keep this TODO for now?

ldematte replied:
It will be already gone in the next PR

mosche reviewed (approved):
LGTM 👍

ldematte commented:
Thanks Moritz!
